src/Func.h

```diff
@@ -375,7 +348,7 @@ class Stage {
     Stage &reorder(const std::vector<VarOrRVar> &vars);
 
     template<typename... Args>
-    HALIDE_NO_USER_CODE_INLINE typename std::enable_if<Internal::all_are_convertible<VarOrRVar, Args...>::value, Stage &>::type
```

Comment:
This can't safely be inlined into user code - it causes useless line numbers in error messages. We want everything in the Halide namespace that could throw a user_error to stay in the Halide namespace, and not get pasted into user code.

Reply:
The error message reporter walks up the stack and reports the line number associated with the first line it finds outside of the Halide:: namespace to help users pinpoint what call into libHalide threw an error. If this method body gets pasted into user code and then the underlying reorder call throws an error then you get an error message blaming Halide.h:34897 instead of your_code.cpp:87

This issue was the origin of HALIDE_NO_USER_CODE_INLINE. For a bunch of functions we wanted inline linkage (because templates), but actually inlining the bodies muddies the boundary between user code and libHalide.

Reply:
Hm, so in that case, why not use HALIDE_NEVER_INLINE?

Reply:
Digging through history:

https://github.com/halide/Halide/commit/d1a40654064f9e893564da0c717873f7b25e5d2f

Reply:
That explains the naming, but it doesn't explain why we are making the sometimes-inline vs never-inline choice.

But at a more meta level, I'm pondering whether this tradeoff is the right decision. The main advantage seems to be that when an error occurs, the user sees a much more helpful line number, so, that's good. But as a downside, it (IMHO) makes it hard to quickly reason about where this code will actually be instantiated -- what we're actually doing here is saying "attempt to inline this if used internally by halide, but *never* inline in user code". That's not unreasonable, but it leaves open the possibility of the internal-to-halide usage not actually being inlined (since it's not a 'force' inline), making it an exported symbol from libHalide... and now we have two versions of it seen by user code (the one from libHalide, plus the one instantiated in user code due to the no-inline directive). The linker will quietly choose one of these, but they could be subtly different due to different compilers being used. I dunno, maybe I'm overthinking this, but quiet unpredictabilities of this sort worry me.

Reply:
I have no problem with just making it never inline.